### PR TITLE
Adding capability to write out Ptrs in JetConstituentSelector

### PR DESCRIPTION
**Pull request by rappoccio**
In order to recluster AK8 jets from MINIAOD (*), we need to write out the jet constituents in a way that can be easily read out. This adds the capability to write out Ptrs and appropriate unpacking of AK8 jets. This should not affect existing workflows. Backports incoming. 

(*) This is often needed for JMAR studies. There will also be an upcoming workflow for JMAR development with nanoaod.

### CODE REVIEW

File: CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc [hunk 1]
```diff
@@ -20,23 +20,29 @@
...
   using JetsOutput = std::vector<T>;
   using ConstituentsOutput = C;
-
+  using ValueType = typename C::value_type;
+  
   JetConstituentSelector(edm::ParameterSet const& params) :
     srcToken_{consumes<edm::View<T>>(params.getParameter<edm::InputTag>("src"))},
-    selector_{params.getParameter<std::string>("cut")}
+    selector_{params.getParameter<std::string>("cut")},
+    unpackAK8_{params.existsAs<bool>("unpackAK8") ? params.getParameter<bool>("unpackAK8") : false}
```
Comment (perrotta) [inline CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc]: Please, add a "false" as default in the fillDescriptions() method, and get rid of the ```existsAs``` here

Comment (rappoccio) [reply CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc]: Won't that affect existing workflows though? If they don't have the parameter they will need to add it. (Or is adding "false" as a default sufficient?)

Comment (perrotta) [reply CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc]: Adding "false" as default in the parameterSetDescription insures that whenever the parameter is missing the default value is assigned.

Which is the same as you are doing now, but a bit cleaner from the point of view of the code.

**Comment by perrotta** [comment]
Adding "false" as default in the parameterSetDescription insures that  
whenever the parameter is missing the default value is assigned.

Which is the same as you are doing now, but a bit cleaner from the  
point of view of the code.

A.

rappoccio <notifications@github.com> ha scritto:

File: CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc (hunk 1 above)
Comment (slava77) [reply CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc]: existsAs is strongly discouraged and exceptions to use it are still possible only for plugins.
This is not the case for this producer.
So, I think that fillDescriptions should be added.

**Comment by davidlange6** [comment]
please test

File: CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc (hunk 1 above)
Comment (rappoccio) [reply CommonTools/RecoAlgos/plugins/JetConstituentSelector.cc]: OK I will fix it. It looks like there are some matrix failures for some reason, I will fix those too.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- The code changes do what said in the description
- The capability to write out the subjets constituents is disabled by default (i.e. standard workflows are not affected) but can be enabled for the possible dedicated use cases

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**